hacspec/src/lib.rs

```diff
@@ -607,59 +609,122 @@ pub fn construct_state(
 }
 
 fn parse_message_1(
-    rcvd_message_1: &BytesMessage1,
-) -> (U8, BytesSupportedSuites, BytesP256ElemLen, U8) {
-    let method = rcvd_message_1[0];
-    // FIXME as we only support a fixed-sized incoming message_1,
-    // we parse directly the selected cipher suite
-    let selected_suite = BytesSupportedSuites::from_slice(rcvd_message_1, 1, 1);
-    let g_x = BytesP256ElemLen::from_slice(rcvd_message_1, 4, P256_ELEM_LEN);
-    let c_i = rcvd_message_1[MESSAGE_1_LEN - 1];
-
-    (method, selected_suite, g_x, c_i)
+    rcvd_message_1: &BufferMessage1,
+) -> Result<(U8, BytesSupportedSuites, BytesP256ElemLen, U8), EDHOCError> {
+    let mut error: EDHOCError = EDHOCError::UnknownError;
+    let method = rcvd_message_1.content[0];
+    let mut g_x: BytesP256ElemLen = BytesP256ElemLen::new();
+    let mut proposed_suites: BytesSuites = BytesSuites::from_public_array([0xff; SUITES_LEN]);
+    let mut proposed_suites_len: usize = 0;
+    let mut selected_suites: BytesSupportedSuites = BytesSupportedSuites::new();
+    let mut c_i = U8(0);
+
+    let result = match (rcvd_message_1.content[1] as U8).declassify() {
+        0x00..=0x17 => {
+            proposed_suites[0] = rcvd_message_1.content[1];
+            Ok(1)
+        }
+        0x18 => {
+            proposed_suites[0] = rcvd_message_1.content[2];
+            Ok(2)
+        }
+        0x80..=0x97 => {
+            let proposed_suites_len: U8 = rcvd_message_1.content[1] - U8(0x80);
+            let proposed_suites_len: usize = proposed_suites_len.declassify().into();
+            for i in 0..proposed_suites_len {
+                proposed_suites[i] = rcvd_message_1.content[2 + i];
+            }
+            Ok(proposed_suites_len + 1)
+        }
+        _ => Err(EDHOCError::ParsingError),
+    };
+
+    if result.is_ok() {
+        proposed_suites_len = result.unwrap();
+        let proposed_suites = proposed_suites.to_public_array();
+        let mut selected_suites_len: usize = 0;
+        for i in 0..EDHOC_SUPPORTED_SUITES.len() {
+            if proposed_suites.contains(&EDHOC_SUPPORTED_SUITES[i].declassify()) {
```

Comment:
This code does not belong in a parsing routine. Parsing routine should just return what was proposed. The invoking code should check whether the proposed suite is indeed supported and act accordingly.